datafusion/physical-plan/src/joins/sort_merge_join.rs

```diff
@@ -989,8 +996,21 @@ impl SMJStream {
                 }
             }
             Ordering::Equal => {
-                if matches!(self.join_type, JoinType::LeftSemi) {
+                if matches!(self.join_type, JoinType::LeftSemi) && self.filter.is_some() {
```

Comment:
You can combine this and below which both are for `JoinType::LeftSemi` under single `if` block.